src/Features/Core/Portable/BraceCompletion/AbstractBraceCompletionService.cs

```diff
@@ -112,20 +112,31 @@ protected virtual Task<bool> IsValidOpenBraceTokenAtPositionAsync(SyntaxToken to
         }
 
         /// <summary>
-        /// Checks that the current position is a valid location.
-        /// Used to determine if overtype should be allowed.
+        /// Returns true when the current position is inside user code (e.g. not strings) and the closing token
+        /// matches the expected closing token for this brace completion service.
+        /// Helper method used by <see cref="AllowOverTypeAsync(BraceCompletionContext, CancellationToken)"/> implementations.
         /// </summary>
-        protected static async Task<bool> CheckCurrentPositionAsync(Document document, int? currentPosition, CancellationToken cancellationToken)
+        protected async Task<bool> AllowOverTypeInUserCodeWithValidClosingTokenAsync(BraceCompletionContext context, CancellationToken cancellationToken)
         {
-            // make sure auto closing is called from a valid position
-            if (currentPosition == null)
+            return await IsCurrentPositionInUserCodeAsync(context.Document, context.CaretLocation, cancellationToken).ConfigureAwait(false)
+                && await CheckClosingTokenKindAsync(context.Document, context.ClosingPoint, cancellationToken).ConfigureAwait(false);
+
+            static async Task<bool> IsCurrentPositionInUserCodeAsync(Document document, int currentPosition, CancellationToken cancellationToken)
             {
-                return false;
+                var tree = await document.GetRequiredSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+                var syntaxFactsService = document.GetRequiredLanguageService<ISyntaxFactsService>();
+                return !syntaxFactsService.IsInNonUserCode(tree, currentPosition, cancellationToken);
             }
```

Comment:
fwiw, this local function doesn't seem to be important :)  could just inline right into the outer method.